linera-core/src/client/mod.rs

```diff
@@ -1132,79 +1185,127 @@ where
-                    // same chain, the call to `receive_certificate` below will download
-                    // the skipped certificate again.
-                    warn!(
-                        "Skipping received certificate from past epoch {:?}",
-                        certificate.value().epoch()
-                    );
-                    new_tracker += 1;
-                }
-                HandleCertificateResult::New => {
-                    certificates.push(certificate);
-                    new_tracker += 1;
+            for certificate in remote_certificates {
+                match self
+                    .check_certificate(max_epoch, &committees, &certificate)
+                    .await?
+                {
+                    HandleCertificateResult::FutureEpoch => {
+                        warn!("Postponing received certificate from {:.8} at height {} from future epoch {}",
+                              chain_id, certificate.value().height(), certificate.value().epoch());
+                        // Stop the synchronization here. Do not increment the tracker further so
+                        // that this certificate can still be downloaded later, once our committee
+                        // is updated.
+                    }
+                    HandleCertificateResult::OldEpoch => {
+                        tracker.push(chain_id, certificate.value().height());
+                        // This epoch is not recognized any more. Let's skip the certificate.
+                        // If a higher block with a recognized epoch comes up later from the
+                        // same chain, the call to `receive_certificate` below will download
+                        // the skipped certificate again.
+                        warn!(
+                            "Skipping received certificate from past epoch {:?}",
+                            certificate.value().epoch()
+                        );
+                    }
+                    HandleCertificateResult::New => {
+                        tracker.push(chain_id, certificate.value().height());
+                        certificates.push(certificate);
+                    }
                 }
             }
         }
+        let new_tracker = tracker.finalize(&remote_log);
         Ok((remote_node.name, new_tracker, certificates))
     }
 
+    // Uses local information (about already-processed blocks) to advance the `block_batch` to a place where only new blocks are left.
+    async fn advance_with_local(
+        &self,
+        chain_id: ChainId,
+        block_batch: &mut Vec<BlockHeight>,
+        tracker: &mut MultichainTracker,
+    ) -> Result<(), NodeError> {
+        let local_response = self.local_query(ChainInfoQuery::new(chain_id)).await?;
+        let local_next = local_response.info.next_block_height;
+
+        tracker.push(chain_id, local_next.saturating_sub(1.into()));
```

Comment:
But if it's `0` we shouldn't remove `0` here.

Reply:
You're right. Done in [c660103](https://github.com/linera-io/linera-protocol/pull/2666/commits/c66010359a10c4462def29e143edf267aec539dc).